proof/refine/ARM/Ipc_R.thy

```diff
@@ -4604,68 +4495,463 @@ lemma maybe_return_sc_weak_valid_sched_action:
+  shows "
+   corres dc (einvs and valid_sched and tcb_at thread and valid_cap cap and ex_nonz_cap_to thread
+              and valid_cap reply_cap
+              and st_tcb_at active thread
+              and not_queued thread and not_in_release_q thread and scheduler_act_not thread
+              and cte_wp_at (\<lambda>c. c = cap.NullCap) (thread, tcb_cnode_index 3)
+              and (\<lambda>s. \<forall>r\<in>zobj_refs reply_cap. ex_nonz_cap_to r s))
+             (invs' and tcb_at' thread and valid_cap' cap' and valid_cap' replyCap)
+             (receive_ipc thread cap isBlocking reply_cap) (receiveIPC thread cap' isBlocking replyCap)"
+  supply if_split [split del]
+  apply (insert assms)
+  apply (rule corres_cross_add_abs_guard[where Q="K (thread \<noteq> idle_thread_ptr)"])
+   apply (clarsimp simp: invs_def valid_state_def valid_pspace_def)
+   apply (frule (1) idle_no_ex_cap)
+   apply (clarsimp simp: valid_idle_def)
+  apply (simp add: receive_ipc_def receiveIPC_def)
+  apply add_sym_refs
+  apply (case_tac cap, simp_all add: isEndpointCap_def)
+  apply (rename_tac word1 word2 right)
+  apply (rule corres_stateAssert_assume)
+   apply (rule stronger_corres_guard_imp)
+   apply (subst receiveIPC_corres_helper)
+     apply (clarsimp simp: receive_ipc_preamble_def[symmetric] receiveIPC_preamble_def[symmetric])
+     apply (rule corres_split_deprecated[OF _ receiveIPC_preamble_corres])
+         apply (rule corres_split_deprecated [OF _ get_ep_corres])
+           apply (rename_tac ep ep')
+           apply (rule corres_split_deprecated [OF _ gbn_corres])
+             apply (rule_tac r'="ntfn_relation" in corres_split_deprecated)
+                apply (rule corres_if)
+                  apply (clarsimp simp: ntfn_relation_def Ipc_A.isActive_def Endpoint_H.isActive_def
+                                 split: Structures_A.ntfn.splits Structures_H.notification.splits)
+                 apply (simp only: )
+                 apply (rule complete_signal_corres)
+                apply (rule corres_split_deprecated[where r'=dc])
+                   apply (rule_tac P="einvs and valid_sched and tcb_at thread
+                                            and ep_at word1 and valid_ep ep and ko_at (Endpoint ep) word1
+                                            and cte_wp_at (\<lambda>c. c = cap.NullCap) (thread, tcb_cnode_index 3)
+                                            and ex_nonz_cap_to thread
+                                            and receive_ipc_preamble_rv reply_cap replyOpt
+                                            and (\<lambda>s. \<forall>r\<in>zobj_refs reply_cap. ex_nonz_cap_to r s)" and
+                                   P'="invs' and tcb_at' thread and ep_at' word1
+                                             and valid_ep' ep'
+                                             and valid_bound_reply' replyOpt"
+                                   in corres_inst)
```

Comment:
The common part is small and a bit awkward, rather than naming the whole small part. I've opted to name each of the highly recurring parts and this make the proofs a bit easier to read (I hope).

Reply:
There’s `corres_inst_add` that lets you specify only some of the conjuncts, which might be useful.